Add AWS cloud provider tags to pre-existing infrastructure

pydctw opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Support writing AWS cloud provider tags on pre-existing infrastructure

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2584

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note:**
```release-note
Support adding AWS cloud provider tags to pre-existing infrastructure
`kubernetes.io/cluster/<name> = shared` and `kubernetes.io/role/elb = 1` for public subnets
`kubernetes.io/cluster/<name> = shared` and`kubernetes.io/role/internal-elb = 1` for private subnets
```

Ankitasw commented:
@pydctw looks like the release note is not written in release note block, that's why bot added `do-not-merge/release-note-label-needed`. You can try to format it, for eg. have a look here https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2718#issue-721321255

pydctw commented:
@scottslowe, could you please review if it meets your requirements?

randomvariable commented:
Do we need to make this opt-in? What if you haven't granted permissions to tag subnets, should we block cluster creation entirely?

randomvariable commented:
Let's also make the release note more specific about what tag is being added.

randomvariable commented:
/priority important-soon

randomvariable commented:
/area networking

scottslowe commented:
@pydctw I'm heading out of town on holiday, and wouldn't be much help in a code review anyway (still learning Golang). Happy to test a pre-release build in my test environment when I get back from holiday, if that would help.

randomvariable commented:
@scottslowe that's fine scott. enjoy your holiday!

pydctw commented:
> Do we need to make this opt-in? What if you haven't granted permissions to tag subnets, should we block cluster creation entirely?

Updated the PR as discussed. If tagging fails for unmanaged subnets, we record an event and proceed to cluster creation. Added a test case named "Unmanaged VPC, 2 existing matching subnets, subnet tagging fails, should succeed" for the case.

sedefsavas commented:
/lgtm

richardcase commented:
Yeah seems good to me as well:

/lgtm

randomvariable commented:
/approve
